File: backend/src/services/stateMachine.ts
import { TripStatus, VehicleStatus, DriverStatus } from "@prisma/client";
import { prisma } from "../lib/prisma.js";
import { AppError } from '../utils/AppError.js';
import { validateVehicleDispatch, validateDriverAssignment } from '../utils/businessRules.js';

export class StateMachineService {
  /**
   * PHASE 2: Dispatch a Trip with full validation rules
   */
  static async dispatchTrip(tripId: string) {
    return await prisma.$transaction(async (tx) => {
      // 1. Fetch the trip along with its current vehicle and driver
      const trip = await tx.trip.findUnique({
        where: { id: tripId },
        include: { vehicle: true, driver: true },
      });

      if (!trip) throw new AppError('Trip not found', 404);
      if (trip.status !== TripStatus.DRAFT) {
        throw new AppError(`Cannot dispatch trip in '${trip.status}' status. Must be DRAFT.`, 400);
      }

      const { vehicle, driver } = trip;

      // 2. Hackathon Business Validation Checks
      validateVehicleDispatch(vehicle.status);
      validateDriverAssignment(driver.status, driver.licenseExpiry);

      if (trip.cargoWeight > vehicle.maxLoadCapacity) {
        throw new AppError(`Cargo weight (${trip.cargoWeight}kg) exceeds vehicle capacity limit (${vehicle.maxLoadCapacity}kg).`, 400);
      }

      // 3. Atomic State Updates
      await tx.vehicle.update({
        where: { id: vehicle.id },
        data: { status: VehicleStatus.ON_TRIP },
      });

      await tx.driver.update({
        where: { id: driver.id },
        data: { status: DriverStatus.ON_TRIP },
      });

      return await tx.trip.update({
        where: { id: tripId },
        data: { status: TripStatus.DISPATCHED },
      });
    });
  }

  /**
   * PHASE 3: Complete Trip (Updates odometer, adds fuel, releases resources)
   */
  static async completeTrip(tripId: string, finalOdometer: number, fuelConsumed: number) {
    return await prisma.$transaction(async (tx) => {
      const trip = await tx.trip.findUnique({ where: { id: tripId } });
      if (!trip || trip.status !== TripStatus.DISPATCHED) {
        throw new AppError('Trip must be in DISPATCHED status to complete.', 400);
      }

      // Release resources back to pool
      await tx.vehicle.update({
        where: { id: trip.vehicleId },
        data: { 
          status: VehicleStatus.AVAILABLE,
          odometer: finalOdometer 
        },
      });

      await tx.driver.update({
        where: { id: trip.driverId },
        data: { status: DriverStatus.AVAILABLE },
      });

      return await tx.trip.update({
        where: { id: tripId },
        data: {
          status: TripStatus.COMPLETED,
          fuelConsumed: fuelConsumed,
        },
      });
    });
  }

  /**
   * PHASE 4: Cancel Trip
   */
  static async cancelTrip(tripId: string) {
    return await prisma.$transaction(async (tx) => {
      const trip = await tx.trip.findUnique({ where: { id: tripId } });
      if (!trip || trip.status === TripStatus.COMPLETED || trip.status === TripStatus.CANCELLED) {
        throw new AppError('Cannot cancel a completed or already cancelled trip.', 400);
      }

      // Reset vehicle and driver back to available if they were dispatched
      if (trip.status === TripStatus.DISPATCHED) {
        await tx.vehicle.update({
          where: { id: trip.vehicleId },
          data: { status: VehicleStatus.AVAILABLE },
        });
        await tx.driver.update({
          where: { id: trip.driverId },
          data: { status: DriverStatus.AVAILABLE },
        });
      }

      return await tx.trip.update({
        where: { id: tripId },
        data: { status: TripStatus.CANCELLED },
      });
    });
  }

  /**
   * PHASE 5: Start & Close Maintenance Log
   */
  static async startMaintenance(vehicleId: string, type: string, anticipatedCost: number) {
    return await prisma.$transaction(async (tx) => {
      await tx.vehicle.update({
        where: { id: vehicleId },
        data: { status: VehicleStatus.IN_SHOP },
      });

      return await tx.maintenanceLog.create({
        data: {
          vehicleId,
          type,
          cost: anticipatedCost,
          startDate: new Date(),
          isActive: true,
        },
      });
    });
  }

  static async closeMaintenance(logId: string, finalCost: number) {
    return await prisma.$transaction(async (tx) => {
      const log = await tx.maintenanceLog.findUnique({ where: { id: logId } });
      if (!log) throw new AppError('Maintenance entry not found', 404);

      await tx.vehicle.update({
        where: { id: log.vehicleId },
        data: { status: VehicleStatus.AVAILABLE },
      });

      return await tx.maintenanceLog.update({
        where: { id: logId },
        data: {
          cost: finalCost,
          endDate: new Date(),
          isActive: false,
        },
      });
    });
  }
}

File: backend/app/services/state_machine.py
from fastapi import HTTPException
from app.models.trip import TripStatus

# Simulating data lookups since other modules are in progress
def mock_get_vehicle(vehicle_id: int):
    # Mocking a valid, active vehicle. Change to return None or specific statuses to test failures.
    return {"id": vehicle_id, "status": "Available", "capacity": 10000, "retired": False}

def mock_get_driver(driver_id: int):
    # Mocking a valid driver with an unexpired license
    return {"id": driver_id, "status": "Available", "license_valid": True}


class StateMachineService:
    @staticmethod
    def dispatch_trip(trip: dict) -> dict:
        """
        Executes the business rules to transition a trip from Draft to Dispatched.
        Updates vehicle and driver statuses inline.
        """
        # 1. Check current trip state
        if trip["status"] != TripStatus.DRAFT:
            raise HTTPException(
                status_code=400, 
                detail=f"Cannot dispatch trip in '{trip['status']}' status. Must be Draft."
            )

        # 2. Fetch vehicle and driver metadata
        vehicle = mock_get_vehicle(trip["vehicle_id"])
        driver = mock_get_driver(trip["driver_id"])

        # 3. Apply Hackathon Business Rules
        if not vehicle or vehicle["retired"]:
            raise HTTPException(status_code=400, detail="Vehicle is retired or does not exist.")
        
        if vehicle["status"] == "In Shop":
            raise HTTPException(status_code=400, detail="Vehicle is currently in maintenance.")
        
        if vehicle["status"] != "Available":
            raise HTTPException(status_code=400, detail="Vehicle is already assigned to another trip.")

        if not driver or not driver["license_valid"]:
            raise HTTPException(status_code=400, detail="Driver does not have a valid license.")

        if driver["status"] != "Available":
            raise HTTPException(status_code=400, detail="Driver is already assigned to another trip.")

        if trip["cargo_weight"] > vehicle["capacity"]:
            raise HTTPException(
                status_code=400, 
                detail=f"Cargo weight ({trip['cargo_weight']}kg) exceeds vehicle capacity ({vehicle['capacity']}kg)."
            )

        # 4. Atomic State Changes if all validations pass
        trip["status"] = TripStatus.DISPATCHED
        vehicle["status"] = "On Trip"
        driver["status"] = "On Trip"

        return trip